hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/HoodieDatasetBulkInsertHelper.scala

```diff
@@ -150,8 +151,8 @@ object HoodieDatasetBulkInsertHelper extends Logging {
       }
 
       writer.getWriteStatuses.asScala.map(_.toWriteStatus).iterator
-    }).collect()
-    table.getContext.parallelize(writeStatuses.toList.asJava)
```

Comment:
I don't think this could be a reason for performance problems. Can you please elaborate what you're trying to achieve here?

cc @boneanxs

Reply:
@alexeykudinkin You can see https://issues.apache.org/jira/browse/HUDI-5327, I introduced the case I encountered in it

Reply:
@alexeykudinkin You can see https://issues.apache.org/jira/browse/HUDI-5327, I introduced the case I encountered in it

Reply:
Hey @Zouxxyy, Thanks for raising this issue!

The reason to collect the data here is that `HoodieData<WriteStatus>` will be used multiple times after `performClustering`, I recall there is an `isEmpty` check could take lots of time, so here we directly convert to a list of WriteStatus, which will reduce the time.

For the second issue, I noticed this and raised a pr to fix it: #7343, will that address your problem? Feel free to review it!

I think `performClusteringWithRecordsRDD` also has the same issue such as using `RDDSpatialCurveSortPartitioner` to optimize data layout, it will call `RDD.isEmpty`, which will raise a new job.

Reply:
@boneanxs 
isEmpty check could take lots of time, I provided a PR to fix it, https://github.com/apache/hudi/pull/7373

Reply:
Yes, we can fix this by directly using `getStat`, but what if `updateIndex` will calculate `writeStatusList` multiple times? If we can directly dereference `RDD<WriteStatus>` to a list of `WriteStatus` at one feasible point(such as `performClusteringWithRecordsAsRow` has already done), we no need to take care of such issue anymore.

As for the parallelism of thread pool could cause the performance issue, I think `performClusteringWithRecordsRDD` also has the same issue. As we might call `partitioner.repartitionRecords`, there could also raise a new job inside the Future thread such as https://github.com/apache/hudi/blob/ea48a85efcf8e331d0cc105d426e830b8bfe5b37/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/execution/bulkinsert/RDDSpatialCurveSortPartitioner.java#L66(check if the RDD is empty or not), or `sortBy` function in `RDDCustomColumnsSortPartitioner`(`sortBy` use `RangePartitoner` which needs to sample the rdd first to decide the ranges, which will also raise a job in the Future)

Reply:
@boneanxs

Reply:
@boneanxs

Reply:
@boneanxs 
First, RDD recalculation is a BUG, we can't add collect here because of it
We need to fix it, with reference to your comments, I modified #7373, try persist writeStatusRDD

For, #7343, you should be right, I overlooked that other operations may also generate a task. However, I'm wondering if it's necessary to specifically set a parameter

Reply:
@boneanxs
For RDD reuse problem, I think we should use persist (fixed https://github.com/apache/hudi/pull/7373) instead of using collcet and creating a new RDD

Reply:
> For, https://github.com/apache/hudi/pull/7343, you should be right, I overlooked that other operations may also generate a job. However, I'm wondering if it's necessary to specifically set a parameter

Very appreciate it if you can review the pr to share your thought, could you please explain more in that pr? :)

Reply:
@Zouxxyy in this case we should actually not be relying on persist as a way to avoid double execution, since persisting is essentially just a caching mechanism (re-using cached blocks on executors) and it'd not be relied upon (it could fail at any point if, for ex, one of the executors fail, making you recompute whole RDD)

Reply:
@alexeykudinkin, ok, WriteStatus has  a large class attributeas `writtenRecords`, as long as collect does not cause OOM